### PR TITLE
[BH-1254] Ensure field slugs and reverse slugs are unique in GraphQL

### PR DESCRIPTION
**Pull request by ChrisWiegman**
## Description

Checks new and edited fields to ensure a slug conflict does not occur when a reverse-relationship is connected. When a duplicate slug is found an error is shown on the appropriate field.

https://wpengine.atlassian.net/browse/BH-1254

## Testing

Appropriate unit tests have been created to determine when a collision should occur and check for the appropriate error.

## Notes

I've done what I believe is extensive testing of new and edited fields to ensure the error is present when needed and only when needed. Please verify the logic in your own testing as, I fully admit, the determination of the error condition was far trickier than I had anticipated.

### CODE REVIEW

**Review by nickcernis — COMMENTED** [review]
I still need to do some more testing, but this is looking good so far, @ChrisWiegman!

Do you see this formatting issue for the error message (Chrome with parcel cache and browser cache cleared), or is it me?

Screenshot 2021-10-05 at 17 44 44

File: includes/rest-api/routes/content-model-field.php [hunk 1]
```diff
@@ -147,6 +148,29 @@ function dispatch_update_content_model_field( WP_REST_Request $request ) {
...
+		isset( $params['enableReverse'] ) &&
+		true === $params['enableReverse'] &&
+		content_model_field_exists(
+			$params['reverseSlug'],
+			$params['id'],
+			$params['reference']
+		)
+	) {
+			return new WP_Error(
+				'wpe_duplicate_field_reverse_slug',
```
Comment (mindctrl) [inline includes/rest-api/routes/content-model-field.php:164]: Most of our WP_Error codes are prefixed with `atlas_content_modeler_` or `atlas-content-modeler-`

**Comment by ChrisWiegman** [comment]
@nickcernis, I've adjusted the message to bring it to one line as I think that looks better than anything I can figure out with CSS.

Screen Shot 2021-10-06 at 12 41 31

File: includes/rest-api/fields.php [hunk 2]
```diff
@@ -100,21 +100,36 @@ function cleanup_detached_relationship_references( string $slug ) {
...
-		if ( $field['slug'] === $slug ) {
-			return true;
+
+		foreach ( $model['fields'] as $field ) {
+			if ( $current_model === $model_id && $field['id'] === $id ) {
+				continue;
+			}
+
+			if ( $model_id === $current_model &&
+			$field['slug'] === $slug ) {
```
Comment (mindctrl) [inline includes/rest-api/fields.php:120]: Seems like this should be idented, but our linter isn't complaining so 🤷

**Review by mindctrl — APPROVED** [review]

**Review by nickcernis — APPROVED** [review]

**Pull request merged**